### PR TITLE
kvclient: don't backoff with unavailable leaseholder

### PR DESCRIPTION
**Pull request by andrewbaptist**
Previously if the leaseholder was unavailable with a ReplicaUnavailableError and we were hitting a NotLeaseHolderError from a different store, we would back off after each failed attempt to the other replicas. This was an unintentional usage of the `InLeaseTransferBackoffs` as it doesn't indicate that the lease is about to move. Instead we should immediately try the next replicas without backoff as long as our cache is up-to-date and the leaseholder is down with a ReplicaUnavailableError.

This PR also updates the test to validate that we are not backing off. Prior to this change we would back off twice on each attempt. A nice benefit is this test runs faster now.

Epic: none

Release note: None

### CODE REVIEW

**Comment by andrewbaptist** [comment]
@arulajmani This was the failure that I was hitting yesterday. After the change to make the NLHE more accurate in detecting whether the leaseholder was updated this test started failing. I'll ping you once this passes CI, but specifically it no longer backs off with a NLHE error on a non-leaseholder replica if the leaseholder hasn't changed.

**Review by arulajmani — COMMENTED** [review]
Nice debugging! 

> Reviewed 1 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119936#-:-NsEQAEw2X7t02rmF4KE:bj9j3ov)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist and @erikgrinaker)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2668 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119936#-NsEPU-c2trdAub6g9TM:-NsEPU-c2trdAub6g9TN:bcbkwk7) ([raw file](https://github.com/cockroachdb/cockroach/blob/5aa27fd8e6823c7d2bae1e918407e155020a6264/pkg/kv/kvclient/kvcoord/dist_sender.go#L2668)):*
> ```Go
> 			case *kvpb.NotLeaseHolderError:
> 				ds.metrics.NotLeaseHolderErrCount.Inc(1)
> 				var lhRUE bool
> ```

nit: 
```go 
var lhRUE bool // lease holder replica unavailable error
```
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2733 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119936#-NsEPrRLDIJsXFSehsux:-NsEPrRLDIJsXFSehsuy:b-ll9bfq) ([raw file](https://github.com/cockroachdb/cockroach/blob/5aa27fd8e6823c7d2bae1e918407e155020a6264/pkg/kv/kvclient/kvcoord/dist_sender.go#L2733)):*
> ```Go
> 					// has not yet applied the new lease.
> 					//
> 					// TODO(arul): The idea here is for the client to not keep sending
> ```

How do you feel about just addressing this TODO instead? I think doing so would have circumvented the issue that you found. It's also a more direct way of deriving `shouldBackoff`, as opposed to the (now) 3 booleans we've got.

**Review by andrewbaptist — COMMENTED** [review]
Thanks - this was a tricky one, especially since it only happened under stressful conditions. The timings were "just right" that the problem didn't normally happen unless it was under stress or stress race. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119936#-:-NsEgHOV-Uais-zX93cA:b-j72zmj)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @erikgrinaker)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2668 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119936#-NsEPU-c2trdAub6g9TM:-NsEhFru2Nh7_zVBNBab:b-bu91xm) ([raw file](https://github.com/cockroachdb/cockroach/blob/5aa27fd8e6823c7d2bae1e918407e155020a6264/pkg/kv/kvclient/kvcoord/dist_sender.go#L2668)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: 
```go 
var lhRUE bool // lease holder replica unavailable error
```
</blockquote></details>

Updated
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2733 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119936#-NsEPrRLDIJsXFSehsux:-NsEhHnE2SQytRTuzTLs:b9yrpb9) ([raw file](https://github.com/cockroachdb/cockroach/blob/5aa27fd8e6823c7d2bae1e918407e155020a6264/pkg/kv/kvclient/kvcoord/dist_sender.go#L2733)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

How do you feel about just addressing this TODO instead? I think doing so would have circumvented the issue that you found. It's also a more direct way of deriving `shouldBackoff`, as opposed to the (now) 3 booleans we've got.
</blockquote></details>

I agree this should be addressed, but let me do this in a follow-up PR. It is hard to reason about all the branches in here now, but this code is also a little fragile with respect to any changes and it is not quite as simple as just changing to track the previous attempt as we have complications with reordering replicas as well. This change is probably simple enough to make, so I might try another PR to do that, but since the partial partition work was blocked on this I wanted to get this through first.

**Review by erikgrinaker — COMMENTED** [review]
I think we should probably be a bit tighter about the condition here -- I'm heading out for the day shortly, but let me have a closer look at this tomorrow.

**Review by arulajmani — COMMENTED** [review]
> I think we should probably be a bit tighter about the condition here -- I'm heading out for the day shortly, but let me have a closer look at this tomorrow.

I think something tighter here would end up addressing the TODO we were discussing below. I'm all for doing so, especially if it isn't too much more effort, but I don't want to block this PR if we've got a ton of partial partitioning work chained behind it. I'm out tomorrow, but I'll let you guys decide on this.

> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119936#-:-NsEl-iH469sTctEZzsu:byuee1e)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2733 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119936#-NsEPrRLDIJsXFSehsux:-NsEi6loEVaUYWpb5pgU:bdjsted) ([raw file](https://github.com/cockroachdb/cockroach/blob/5aa27fd8e6823c7d2bae1e918407e155020a6264/pkg/kv/kvclient/kvcoord/dist_sender.go#L2733)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

I agree this should be addressed, but let me do this in a follow-up PR. It is hard to reason about all the branches in here now, but this code is also a little fragile with respect to any changes and it is not quite as simple as just changing to track the previous attempt as we have complications with reordering replicas as well. This change is probably simple enough to make, so I might try another PR to do that, but since the partial partition work was blocked on this I wanted to get this through first.
</blockquote></details>

Okay, let's do it in a followup. 

In that case, let's update the comment above to talk about the `!lhRUE` condition as well. Something like "we're not going to retry the same leaseholder because it's unavailable, so there's no need to backoff"
___
*[`pkg/kv/kvserver/client_replica_circuit_breaker_test.go` line 799 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119936#-NsEkvi10xqdDFRbjAD8:-NsEkvi10xqdDFRbjAD9:bcbfss5) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8fd185492fb41ab61f3376271b31e86d874386/pkg/kv/kvserver/client_replica_circuit_breaker_test.go#L799)):*
> ```Go
> 			initialBackoff := backoffMetric.Count()
> 			err := db.Put(ctx, key, value)
> 			require.EqualValues(t, 0, backoffMetric.Count()-initialBackoff)
> ```

Should we add a comment explaining this?

**Review by andrewbaptist — COMMENTED** [review]
TFTR - I can chat with Erik in the morning if he thinks we should change this.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119936#-:-NsFW_GtCKJSDDf5BXMi:b-9zyrg2)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @erikgrinaker)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2733 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119936#-NsEPrRLDIJsXFSehsux:-NsFVzU02UXgKENXiIVY:b-kv93sp) ([raw file](https://github.com/cockroachdb/cockroach/blob/5aa27fd8e6823c7d2bae1e918407e155020a6264/pkg/kv/kvclient/kvcoord/dist_sender.go#L2733)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Okay, let's do it in a followup. 

In that case, let's update the comment above to talk about the `!lhRUE` condition as well. Something like "we're not going to retry the same leaseholder because it's unavailable, so there's no need to backoff"
</blockquote></details>

Ack - Added a comment and  I'll wait for @erikgrinaker to verify and will work on a change for this in parallel if we decide to change it. I think this change is good regardless of partial partitions because we have retries at the outer layer (sendPartialBatch) that will be called in the case of all replicas being unavailable or unable to acquire the lease. Its unclear why backing off between individual replicas (previous behavior) would ever be preferable to backing off after we try all the replicas.
___
*[`pkg/kv/kvserver/client_replica_circuit_breaker_test.go` line 799 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119936#-NsEkvi10xqdDFRbjAD8:-NsFW4U78_UXu0Tu0Of-:b-8b11j9) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8fd185492fb41ab61f3376271b31e86d874386/pkg/kv/kvserver/client_replica_circuit_breaker_test.go#L799)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should we add a comment explaining this?
</blockquote></details>

Added

**Review by erikgrinaker — APPROVED** [review]
Thanks for the investigation and fix! This was an oversight in #118737.

> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119936#-:-NsIswUYDIxoXGr2sM-F:bw6rsm5)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist and @arulajmani)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2733 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119936#-NsEPrRLDIJsXFSehsux:-NsIqMy84Y1f8OUg1Cw9:b-rz6ykd) ([raw file](https://github.com/cockroachdb/cockroach/blob/5aa27fd8e6823c7d2bae1e918407e155020a6264/pkg/kv/kvclient/kvcoord/dist_sender.go#L2733)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Ack - Added a comment and  I'll wait for @erikgrinaker to verify and will work on a change for this in parallel if we decide to change it. I think this change is good regardless of partial partitions because we have retries at the outer layer (sendPartialBatch) that will be called in the case of all replicas being unavailable or unable to acquire the lease. Its unclear why backing off between individual replicas (previous behavior) would ever be preferable to backing off after we try all the replicas.
</blockquote></details>

I think what we really want to check here is the `moved` from where we moved the leaseholder to the front of the transport. In other words, if we're trying the same leaseholder again next, then we should backoff before trying it.

The conditions here are trying to express exactly that, but they do so incompletely -- as another example, if we bail out due to `sameReplicaRetryLimit` this will end up incorrectly adding a backoff afterwards.

I have a mild preference for adding a `movedLeaseholderToFront` boolean or something that we set if we moved a new leaseholder to the front, and making the condition here `movedLeaseholderToFront && !updatedLeaseholder && !intentionallySentToFollower`, but I'll leave that to you to decide. We can also address that in #119747, as a more general cleanup.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2733 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119936#-NsIq6dL3CCwz7a7ANfq:-NsIq6dL3CCwz7a7ANfr:b-a8d6v2) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb71f52c11e030d54ee6a82ba0ea8e9cf469f05c/pkg/kv/kvclient/kvcoord/dist_sender.go#L2733)):*
> ```Go
> 					// has not yet applied the new lease.
> 					//
> 					// If the leaseholder is unavailable (!lhRUE), we don't backoff since
> ```

`!lhRUE` should be `lhRUE` (we don't backoff if this is true).

**Comment by andrewbaptist** [comment]
bors r=erikgrinaker 

TFTR - I'm changing the condition again as part of #119791 - so I left the more complex condition. The condition is not exactly the same as the one above, and we could potentially tighten the backoff conditions even more, but the remaining exceptions are pretty minor and I would rather address the more major one first since that is what blocks the partial partition work. I have a much larger refactor planned of this code which I will likely work on during 24.2 to really improve the behavior of the DistSender in the case of unavailability, but that is going to require a much more extensive test suite, so for now I'm just patching holes that cause actual problems.

**Pull request merged**
